### ISSUE TITLE
"PasswordAuthentication yes" required for user/pass login to OpenSSH server 

### ISSUE THREAD

**Issue by yolabingo** [phpseclib/phpseclib#1210]
When connecting with `phpseclib\Net\SFTP->login('user','password')` to an OpenSSH server that has the following settings in /etc/ssh/sshd_config
```
# To disable tunneled clear text passwords, change to no here!
PasswordAuthentication no
ChallengeResponseAuthentication yes
```
phpseclib will connect but never complete auth, and exit with:
```
Expected SSH_MSG_KEXINIT in phpseclib/phpseclib/Net/SSH2.php on line 1467
Connection closed by server in phpseclib/phpseclib/Net/SSH2.php on line 3586
```

These are the default settings in sshd_config for Gentoo linux. I have not encountered another SSH client that has issues connecting to this OpenSSH server with the default Gentoo settings. 
Setting `PasswordAuthentication yes` in sshd_config allows phpseclib to connect.

Fedora, Debian and Ubuntu have a default of `PasswordAuthentication yes` for the record.

I don't know that this is a bug, but I'm posting as these OpenSSH server settings are not unheard of so others may encounter this issue.

**Comment by terrafrost** [comment]
I connected to an SSH server that had the same configuration yours did without issue. More specifically, I used [Vagrant](https://en.wikipedia.org/wiki/Vagrant_(software)) to create a VM with an altered /etc/ssh/sshd_config so that my results could be reproduced. Here's my Vagrantfile:

```
Vagrant.configure("2") do |config|
	config.vm.box = "laravel/homestead"
	config.vm.provision "shell", inline: <<-SHELL
		sed -i 's/ChallengeResponseAuthentication no/ChallengeResponseAuthentication yes/g' /etc/ssh/sshd_config
		sed -i 's/#PasswordAuthentication yes/PasswordAuthentication no/g' /etc/ssh/sshd_config
		service ssh restart
	SHELL
end
```

My PHP code:

```php
<?php
require __DIR__ . '/vendor/autoload.php';

use phpseclib\Net\SFTP;

$sftp = new SFTP('127.0.0.1', 2222);
$sftp->login('vagrant', 'vagrant');

print_r($sftp->nlist());
```

So clearly the configuration you described is not an issue. For issues like the one you're having logs are most helpful. Let's enable logging to the above code to see what's happening. To do so we'll add `define('NET_SSH2_LOGGING', 2);` to the top of the file and `echo $sftp->getLog();` to the end.

Here's a snippet from the logs that I got:

```
-> NET_SSH2_MSG_SERVICE_REQUEST (since last: 0.0018, network: 0s)
00000000  00:00:00:0c:73:73:68:2d:75:73:65:72:61:75:74:68  ....ssh-userauth

<- NET_SSH2_MSG_SERVICE_ACCEPT (since last: 0.0004, network: 0.0003s)
00000000  00:00:00:0c:73:73:68:2d:75:73:65:72:61:75:74:68  ....ssh-userauth

-> NET_SSH2_MSG_USERAUTH_REQUEST (since last: 0.0001, network: 0s)
00000000  00:00:00:08:75:73:65:72:6e:61:6d:65:00:00:00:0e  ....username....
00000010  73:73:68:2d:63:6f:6e:6e:65:63:74:69:6f:6e:00:00  ssh-connection..
00000020  00:08:70:61:73:73:77:6f:72:64:00:00:00:00:08:70  ..password.....p
00000030  61:73:73:77:6f:72:64                             assword

<- NET_SSH2_MSG_USERAUTH_FAILURE (since last: 0.0015, network: 0.0014s)
00000000  00:00:00:1e:70:75:62:6c:69:63:6b:65:79:2c:6b:65  ....publickey,ke
00000010  79:62:6f:61:72:64:2d:69:6e:74:65:72:61:63:74:69  yboard-interacti
00000020  76:65:00                                         ve.

-> NET_SSH2_MSG_USERAUTH_REQUEST (since last: 0.0001, network: 0s)
00000000  00:00:00:07:76:61:67:72:61:6e:74:00:00:00:0e:73  ....vagrant....s
00000010  73:68:2d:63:6f:6e:6e:65:63:74:69:6f:6e:00:00:00  sh-connection...
00000020  14:6b:65:79:62:6f:61:72:64:2d:69:6e:74:65:72:61  .keyboard-intera
00000030  63:74:69:76:65:00:00:00:00:00:00:00:00           ctive........

<- NET_SSH2_MSG_USERAUTH_INFO_REQUEST (60) (since last: 0.0007, network: 0.0007s)
00000000  00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:01  ................
00000010  00:00:00:0a:50:61:73:73:77:6f:72:64:3a:20:00     ....Password: .

-> NET_SSH2_MSG_USERAUTH_INFO_RESPONSE (61) (since last: 0.0001, network: 0s)
00000000  00:00:00:01:00:00:00:0c:64:75:6d:6d:79:2d:61:6e  ........dummy-an
00000010  73:77:65:72                                      swer

<- NET_SSH2_MSG_USERAUTH_INFO_REQUEST (60) (since last: 0.0059, network: 0.0058s)
00000000  00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00  ................

-> NET_SSH2_MSG_USERAUTH_INFO_RESPONSE (61) (since last: 0.0004, network: 0.0001s)
00000000  00:00:00:00                                      ....

<- NET_SSH2_MSG_USERAUTH_SUCCESS (since last: 0.0017, network: 0.0016s)
```

So you can see that it tries password auth first. The server responds by saying that it only supports keyboard-interactive auth (ie. "ChallengeResponseAuthentication") so then phpseclib tries that. The prompt it gets from the server is "Password:" and phpseclib then responds with the password (denoted in the logs as "dummy-answer" since the logs filter out the password).

The error message you're getting is `Expected SSH_MSG_KEXINIT`. That error message precedes the authentication phase. If the server and the client can't settle on the algorithms used then the password would be sent in clear text. And phpseclib can't settle on the algorithms used by the server because I guess the server isn't sending them.

idk what version of phpseclib you're using but maybe you're not using the latest version. phpseclib 1.0.7 and 2.0.6 both expected the server to send the KEX packet first. #1162 elaborates. If you are using the latest version then maybe restoring the old behavior by doing `$ssh->sendKEXINITLast();` will do the trick for you.

If neither of those works than I'd say post the server logs. As previously noted, you can do so by doing `define('NET_SSH2_LOGGING', 2);` to the top of the file and `echo $sftp->getLog();` to the end.

**Comment by terrafrost** [comment]
TLDR try the latest version if you're not already. If you are using the latest version and still having the issueget me the SSH logs, which you can do by doing `define('NET_SSH2_LOGGING', 2);` at the top and `echo $sftp->getLog();` at the end.

Thanks!

**Comment by yolabingo** [comment]
Thanks much for this very detailed response.
